File: opspilot-enterprise/packages/shared-types/src/incident.ts
export type IncidentStatus =
  | "new"
  | "analyzing"
  | "pending_action"
  | "resolved"
  | "archived";

export type SeverityLevel = "critical" | "high" | "medium" | "low" | "info";

export interface Incident {
  id: string;
  title: string;
  status: IncidentStatus;
  severity: SeverityLevel;
  source: string;
  source_type: string;
  affected_objects: AffectedObject[];
  first_seen_at: string;
  last_updated_at: string;
  owner: string | null;
  ai_analysis_triggered: boolean;
  root_cause?: RootCause;
  root_cause_candidates: RootCauseCandidate[];
  hypotheses?: Hypothesis[];
  winning_hypothesis?: Hypothesis;
  counter_evidence_result?: CounterEvidenceResult;
  conclusion_status?: "confirmed" | "probable" | "insufficient_evidence" | "contradicted";
  evidence_sufficiency?: EvidenceSufficiency;
  contradictions?: Array<{ type: string; summary: string; evidence_refs?: string[]; severity?: string }>;
  recommended_actions: string[];
  evidence_refs: string[];
  analysis?: IncidentAnalysis;
  summary: string;
}

export interface AffectedObject {
  object_type: string;
  object_id: string;
  object_name: string;
}

export interface RootCauseCandidate {
  id: string;
  description: string;
  confidence: number;
  evidence_refs: string[];
  category: string;
}

export interface RootCause {
  summary: string;
  confidence: number;
  evidence_refs: string[];
}

export interface EvidenceSufficiency {
  required_evidence_types: string[];
  present_evidence_types: string[];
  missing_critical_evidence: string[];
  sufficiency_score: number;
  freshness_score: number;
}

export interface Hypothesis {
  id: string;
  summary: string;
  category: string;
  confidence: number;
  support_evidence_refs: string[];
  counter_evidence_refs: string[];
  missing_evidence: string[];
  status: "candidate" | "confirmed" | "probable" | "refuted" | "inconclusive";
  why: string;
}

export interface CounterEvidenceResult {
  status: "refuted" | "not_refuted" | "inconclusive";
  checked_hypothesis_id?: string | null;
  summary: string;
  evidence_refs: string[];
}

export type AnalysisStatus = "idle" | "running" | "completed" | "failed";

export interface IncidentAnalysisStep {
  round: number;
  stage: string;
  goal?: string;
  tool_name?: string;
  selected_tools?: string[];
  input_summary?: string;
  output_summary?: string;
  evidence_found?: string[];
  evidence_missing?: string[];
  contradictions?: string[];
  finding: string;
  decision: string;
  why?: string;
  timestamp: string;
  status: "success" | "failed" | "running";
}

export interface IncidentAnalysis {
  status: AnalysisStatus;
  round: number;
  max_rounds: number;
  started_at?: string;
  updated_at?: string;
  elapsed_ms?: number;
  final_conclusion?: string;
  recommended_actions?: string[];
  analysis_process: IncidentAnalysisStep[];
  next_decision?: string;
}

export interface IncidentTimelineEntry {
  timestamp: string;
  type: "event" | "analysis" | "notification" | "action" | "resolution";
  summary: string;
  agent?: string;
  detail?: string;
}

File: packages/shared-schema/src/opspilot_schema/incident.py
from __future__ import annotations
from typing import Literal
from pydantic import BaseModel


class AffectedObject(BaseModel):
    object_type: str
    object_id: str
    object_name: str


class RootCauseCandidate(BaseModel):
    id: str
    description: str
    confidence: float
    evidence_refs: list[str] = []
    category: str


class RootCause(BaseModel):
    summary: str
    confidence: float
    evidence_refs: list[str] = []


class EvidenceSufficiency(BaseModel):
    required_evidence_types: list[str] = []
    present_evidence_types: list[str] = []
    missing_critical_evidence: list[str] = []
    sufficiency_score: float = 0.0
    freshness_score: float = 0.0


class Hypothesis(BaseModel):
    id: str
    summary: str
    category: str
    confidence: float
    support_evidence_refs: list[str] = []
    counter_evidence_refs: list[str] = []
    missing_evidence: list[str] = []
    status: Literal["candidate", "confirmed", "probable", "refuted", "inconclusive"] = "candidate"
    why: str = ""


class CounterEvidenceResult(BaseModel):
    status: Literal["refuted", "not_refuted", "inconclusive"]
    checked_hypothesis_id: str | None = None
    summary: str
    evidence_refs: list[str] = []


class Incident(BaseModel):
    id: str
    title: str
    status: Literal["new", "analyzing", "pending_action", "resolved", "archived"]
    severity: Literal["critical", "high", "medium", "low", "info"]
    source: str
    source_type: str
    affected_objects: list[AffectedObject] = []
    first_seen_at: str
    last_updated_at: str
    owner: str | None = None
    ai_analysis_triggered: bool = False
    root_cause: RootCause | None = None
    root_cause_candidates: list[RootCauseCandidate] = []
    hypotheses: list[Hypothesis] = []
    winning_hypothesis: Hypothesis | None = None
    counter_evidence_result: CounterEvidenceResult | None = None
    conclusion_status: Literal["confirmed", "probable", "insufficient_evidence", "contradicted"] | None = None
    evidence_sufficiency: EvidenceSufficiency | None = None
    contradictions: list[dict] = []
    recommended_actions: list[str] = []
    evidence_refs: list[str] = []
    summary: str


class IncidentTimelineEntry(BaseModel):
    timestamp: str
    type: Literal["event", "analysis", "notification", "action", "resolution"]
    summary: str
    agent: str | None = None
    detail: str | None = None
